capsules/src/process_console.rs

```diff
@@ -120,6 +144,64 @@ pub struct ProcessConsole<'a, A: Alarm<'a>, C: ProcessManagementCapability> {
     capability: C,
 }
 
+#[derive(Copy, Clone)]
+pub struct Command {
+    buf: [u8; COMMAND_BUF_LEN],
+    len: usize,
+}
+
+impl Command {
+    /// Write the buffer with the provided data.
+    /// If the provided data's length is smaller than the buffer length,
+    /// the left over bytes are not modified due to '\0' termination.
+    fn write(&mut self, buf: &[u8; COMMAND_BUF_LEN]) {
+        self.len = buf
+            .iter()
+            .position(|a| *a == EOL)
+            .unwrap_or(COMMAND_BUF_LEN);
+
+        (&mut self.buf).copy_from_slice(buf);
+    }
+
+    fn insert_byte(&mut self, byte: u8) {
+        if self.len < COMMAND_BUF_LEN {
+            self.buf[self.len] = byte;
+            self.len = self.len + 1;
+        }
+    }
+
+    fn delete_last_byte(&mut self) {
+        if self.len > 0 {
+            self.len -= 1;
+            self.buf[self.len] = EOL;
+        }
+    }
+
+    fn clear(&mut self) {
+        self.buf.iter_mut().for_each(|x| *x = EOL);
+        self.len = 0;
+    }
+}
+
+impl Default for Command {
+    fn default() -> Self {
+        Command {
+            buf: [EOL; COMMAND_BUF_LEN],
+            len: 0,
+        }
+    }
+}
+
+impl PartialEq<[u8; COMMAND_BUF_LEN]> for Command {
+    fn eq(&self, other_buf: &[u8; COMMAND_BUF_LEN]) -> bool {
+        self.buf
+            .iter()
+            .zip(other_buf.iter())
+            .take_while(|(a, b)| **a != EOL || **b != EOL)
+            .all(|(a, b)| *a == *b)
```

Comment:
Won't this not work if one is longer then the other?

For example if the buffers are: "abc" and "abcde" won't zip just iterate "abc" and they all match?

Reply:
Cheked once again on the board and seems to work as expected, also have tested it on your example and seems to work.

The `take_while` will stop until both bytes from buffer are set to `EOL`.
In other words will stop when will encounter a `(EOL, EOL)` pair.
So for your example we have the following buffers:
```text
b1 = ['a', 'b', 'c', EOL, EOL, EOL, ....];
b2 = ['a', 'b', 'c', 'd', 'e', EOL, EOL, EOL, ...];
```

when comparing we will have the following pairs:
```text
zip = [
('a', 'a'),
('b', 'b'),
('c', 'c'),
(EOL, 'd'),
(EOL, 'e')
]
```
Which will evaluate as false.

Instead if we put on the take_while predicate something like this `|(a, b)| **a != EOL && **b != EOL` it would generate the behaviour that you were talking about.

Reply:
Ah! Your right. While I was looking at this I forgot they are fixed length buffers :facepalm: